torch/onnx/_internal/diagnostics/infra/_infra.py

```diff
@@ -0,0 +1,349 @@
+
+
+@dataclasses.dataclass
+class Diagnostic:
+    rule: Rule
+    level: Level
+    message_args: Optional[Tuple[Any, ...]]
+    locations: List[Location] = dataclasses.field(default_factory=list)
+    stacks: List[Stack] = dataclasses.field(default_factory=list)
+    additional_message: Optional[str] = None
+    tags: List[Tag] = dataclasses.field(default_factory=list)
+
+    def sarif(self) -> sarif.Result:
+        """Returns the SARIF Result representation of this diagnostic."""
+        if self.message_args is None:
+            self.message_args = tuple()
+        message = self.rule.message_default_template.format(*self.message_args)
+        if self.additional_message is not None:
+            message = f"{message}\n{self.additional_message}"
+        sarif_result = sarif.Result(
+            message=sarif.Message(text=message),
+            level=self.level.value,
+            rule_id=self.rule.id,
+        )
+        sarif_result.locations = [location.sarif() for location in self.locations]
+        sarif_result.stacks = [stack.sarif() for stack in self.stacks]
+        sarif_result.properties = sarif.PropertyBag(
+            tags=[tag.value for tag in self.tags]
+        )
+        return sarif_result
+
+    def with_location(self: _Diagnostic, location: Location) -> _Diagnostic:
+        """Adds a location to the diagnostic."""
+        self.locations.append(location)
+        return self
+
+    def with_stack(self: _Diagnostic, stack: Stack) -> _Diagnostic:
+        """Adds a stack to the diagnostic."""
+        self.stacks.append(stack)
+        return self
+
+    def with_additional_message(self: _Diagnostic, message: str) -> _Diagnostic:
+        """Adds an additional message to the diagnostic."""
+        if self.additional_message is None:
+            self.additional_message = message
+        else:
+            self.additional_message = f"{self.additional_message}\n{message}"
+        return self
+
+
+@dataclasses.dataclass
+class RuleCollection:
+    _rule_id_name_set: FrozenSet[Tuple[str, str]] = dataclasses.field(init=False)
+
+    def __post_init__(self) -> None:
+        self._rule_id_name_set = frozenset(
+            {
+                (field.default.id, field.default.name)
+                for field in dataclasses.fields(self)
+                if isinstance(field.default, Rule)
+            }
+        )
+
+    def __contains__(self, rule: Rule) -> bool:
+        """Checks if the rule is in the collection."""
+        return (rule.id, rule.name) in self._rule_id_name_set
+
+    @classmethod
+    def custom_collection_from_list(
+        cls, new_collection_class_name: str, rules: Sequence[Rule]
+    ) -> RuleCollection:
+        """Creates a custom class inherited from RuleCollection with the list of rules."""
+        return dataclasses.make_dataclass(
+            new_collection_class_name,
+            [
+                (
+                    formatter._kebab_case_to_snake_case(rule.name),
```

Comment:
Looks like `_kebab_case_to_snake_case` is used outside of `formatter`. I would rename it to `kebab_case_to_snake_case` to make it public